File: bindings/kotlin/scp-kt/src/test/kotlin/works/limn/scp/CaveatsRoundtripTest.kt
// CaveatsRoundtripTest.kt — SCP-OUT-023 AC-7 conformance for the Kotlin SDK.
//
// Mirrors `bindings/python/tests/test_caveats_roundtrip.py`. Builds an
// `InvocationCaveats` JSON wire object, mints a UCAN through the real
// UniFFI bridge (`uniffi.scp.ucanMint(..., caveatsJson)`), decodes the
// returned JWT's payload segment, and asserts every populated caveat
// field surfaces in `payload.nb` byte-for-byte.
//
// The test skips cleanly when the UniFFI-generated native library is not
// available (mirrors `pytest.importorskip` in the Python conformance test
// and the `assumeTrue(nativeAvailable, ...)` pattern in `RealFFITest`).
//
// Provenance:
//   - .docs/prds/outlet.json — SCP-OUT-023 AC-7
//   - .docs/specs/07-trust-validation-and-capabilities.md §7.3.8
//   - bindings/python/tests/test_caveats_roundtrip.py (reference)

package works.limn.scp

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Assumptions.assumeTrue
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.util.Base64
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

@Suppress("TooManyFunctions")
@OptIn(ExperimentalCoroutinesApi::class)
class CaveatsRoundtripTest {
    companion object {
        private var nativeAvailable = false
        private var skipReason = ""
        private var relayHandle: Any? = null

        @JvmStatic
        @BeforeAll
        fun checkNativeLibraryAndStartRelay() {
            try {
                Class.forName("uniffi.scp.ScpKt")
                nativeAvailable = true
                kotlinx.coroutines.runBlocking {
                    val handle = uniffi.scp.relayStartInMemory()
                    relayHandle = handle
                    val bootstrap = uniffi.scp.identityCreate("in_memory")
                    uniffi.scp.configureRelayTransport(
                        relayUrl = handle.relayUrl(),
                        localDid = bootstrap.did(),
                    )
                }
            } catch (e: ClassNotFoundException) {
                skipReason = "UniFFI generated bindings not available: ${e.message}"
            } catch (e: UnsatisfiedLinkError) {
                skipReason = "Native library link error: ${e.message}"
            } catch (e: ExceptionInInitializerError) {
                skipReason = "Native library init error: ${e.cause?.message ?: e.message}"
            } catch (e: NoClassDefFoundError) {
                skipReason = "Native library class not found: ${e.message}"
            }
        }

        @JvmStatic
        @AfterAll
        fun shutdownRelay() {
            (relayHandle as? uniffi.scp.RelayHandle)?.shutdown()
        }

        /** Decodes the JWT payload (middle base64url segment) into a JsonObject. */
        fun decodeJwtPayload(encoded: String): JsonObject {
            val parts = encoded.split(".")
            require(parts.size == 3) { "expected 3 JWT segments, got ${parts.size}" }
            val padded = parts[1] + "=".repeat((4 - parts[1].length % 4) % 4)
            val bytes = Base64.getUrlDecoder().decode(padded)
            return Json.parseToJsonElement(String(bytes, Charsets.UTF_8)).jsonObject
        }

        /** Default cross-delegation context params with a relaxed ceiling. */
        private fun defaultParams(ceiling: List<String>) =
            uniffi.scp.ContextParams(
                mode = uniffi.scp.ContextMode.ENCRYPTED,
                ceiling = ceiling,
                ceilingPolicy = uniffi.scp.CeilingPolicy.IMMUTABLE,
                governance = uniffi.scp.GovernanceModel.SINGLE_ADMIN,
                memoryScope = uniffi.scp.MemoryScope.EPHEMERAL,
                ttlSeconds = 0uL,
                promotable = false,
                minProtocolVersion = 0u,
            )

        /** Builds the 8-budgeted-fields + originKind=Action wire JSON for caveats. */
        fun primaryCaveatsJson(): String =
            buildJsonObject {
                put("amountMaxPerCall", 100)
                put("amountMaxCumulative", 1000)
                put("validFrom", 1_700_000_000)
                put("validUntil", 1_700_003_600)
                put("maxCalls", 42)
                putJsonObject("rateWindow") {
                    put("max", 1)
                    put("windowSecs", 60)
                }
                putJsonArray("allowedAdapters") {
                    add("native")
                    add("openai-compatible")
                }
                putJsonArray("allowedTargetDids") {
                    add("did:dht:zMember")
                    add("did:dht:zOther")
                }
                put("originKind", "Action")
            }.toString()

        /** Builds the companion (hoursOfDay/daysOfWeek/inputSchema/originKind=Query) wire JSON. */
        fun companionCaveatsJson(): String =
            buildJsonObject {
                put("hoursOfDay", 0x00FF_FFFF)
                put("daysOfWeek", 0x7F)
                putJsonObject("inputSchema") {
                    put("type", "object")
                    putJsonObject("properties") {
                        putJsonObject("x") { put("type", "number") }
                    }
                    putJsonArray("required") { add("x") }
                }
                put("originKind", "Query")
            }.toString()

        /** Asserts the companion mint's `nb` field carries every populated caveat. */
        fun assertCompanionNb(payloadNb: JsonObject) {
            assertEquals(0x00FF_FFFF, payloadNb["hoursOfDay"]?.jsonPrimitive?.int)
            assertEquals(0x7F, payloadNb["daysOfWeek"]?.jsonPrimitive?.int)
            val schema = payloadNb["inputSchema"]?.jsonObject
            assertNotNull(schema, "inputSchema must be present")
            assertEquals("object", schema["type"]?.jsonPrimitive?.content)
            val required = schema["required"]?.jsonArray
            assertNotNull(required, "inputSchema.required must be present")
            assertEquals(listOf("x"), required.map { it.jsonPrimitive.content })
            assertEquals("Query", payloadNb["originKind"]?.jsonPrimitive?.content)
            for (absent in listOf(
                "amountMaxPerCall",
                "amountMaxCumulative",
                "validFrom",
                "validUntil",
                "maxCalls",
                "rateWindow",
                "allowedAdapters",
                "allowedTargetDids",
            )) {
                assertFalse(payloadNb.containsKey(absent), "field $absent must be omitted from nb")
            }
        }

        /** Builds a 9-populated-field caveats wire JSON that exceeds MAX_POPULATED_CAVEATS. */
        fun overCapCaveatsJson(): String =
            buildJsonObject {
                put("amountMaxPerCall", 1)
                put("amountMaxCumulative", 2)
                put("validFrom", 3)
                put("validUntil", 4)
                put("hoursOfDay", 0x00FF_FFFF)
                put("daysOfWeek", 0x7F)
                put("maxCalls", 5)
                putJsonObject("rateWindow") {
                    put("max", 1)
                    put("windowSecs", 60)
                }
                putJsonObject("inputSchema") { put("type", "object") } // 9th field
            }.toString()

        /** Asserts the primary mint's `nb` field carries every populated caveat. */
        fun assertPrimaryNb(payloadNb: JsonObject) {
            assertEquals(100, payloadNb["amountMaxPerCall"]?.jsonPrimitive?.int)
            assertEquals(1000, payloadNb["amountMaxCumulative"]?.jsonPrimitive?.int)
            assertEquals(1_700_000_000L, payloadNb["validFrom"]?.jsonPrimitive?.long)
            assertEquals(1_700_003_600L, payloadNb["validUntil"]?.jsonPrimitive?.long)
            assertEquals(42, payloadNb["maxCalls"]?.jsonPrimitive?.int)

            val rateWindow = payloadNb["rateWindow"]?.jsonObject
            assertNotNull(rateWindow, "rateWindow must be present and a JSON object")
            assertEquals(1, rateWindow["max"]?.jsonPrimitive?.int)
            assertEquals(60, rateWindow["windowSecs"]?.jsonPrimitive?.int)

            val adapters = payloadNb["allowedAdapters"]?.jsonArray
            assertNotNull(adapters, "allowedAdapters must be present")
            assertEquals(listOf("native", "openai-compatible"), adapters.map { it.jsonPrimitive.content })

            val targets = payloadNb["allowedTargetDids"]?.jsonArray
            assertNotNull(targets, "allowedTargetDids must be present")
            assertEquals(
                listOf("did:dht:zMember", "did:dht:zOther"),
                targets.map { it.jsonPrimitive.content },
            )

            assertEquals("Action", payloadNb["originKind"]?.jsonPrimitive?.content)

            // SCP-OUT-018 `skip_serializing_if = "Option::is_none"`: omitted
            // SDK fields must not appear in `nb`, never as null.
            assertFalse(payloadNb.containsKey("hoursOfDay"), "absent field must be omitted")
            assertFalse(payloadNb.containsKey("daysOfWeek"), "absent field must be omitted")
            assertFalse(payloadNb.containsKey("inputSchema"), "absent field must be omitted")
        }
    }

    // §7.3.8 mint-limit: at most MAX_POPULATED_CAVEATS = 8 non-origin_kind
    // fields populated per envelope (origin_kind is structural and exempt).
    // This first test populates 8 budgeted fields + originKind — the
    // maximum mintable shape — and asserts every field round-trips through
    // the JWT `nb` field.
    @Test
    fun `8 budgeted caveats plus originKind round-trip via UCAN nb (SCP-OUT-023 AC-7)`() =
        runTest {
            assumeTrue(nativeAvailable, skipReason)

            // Cross-delegation: admin mints for member (avoids ADR-039 self-mint).
            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val admin = uniffi.scp.identityCreate("in_memory")
            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val member = uniffi.scp.identityCreate("in_memory")
            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val handle =
                uniffi.scp.contextCreate(
                    admin,
                    defaultParams(listOf("messages:read", "messages:write")),
                )

            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val token =
                uniffi.scp.ucanMint(
                    handle,
                    member.did(),
                    listOf("messages:write"),
                    null,
                    primaryCaveatsJson(),
                )

            val encoded = token.encoded()
            assertTrue(encoded.isNotEmpty(), "token.encoded() must be non-empty")

            val payloadNb =
                decodeJwtPayload(encoded)["nb"]?.jsonObject
                    ?: error("JWT payload missing `nb` field")
            assertPrimaryNb(payloadNb)
        }

    // Companion mint covering the three fields the primary test omitted
    // (hoursOfDay, daysOfWeek, inputSchema) plus originKind=Query. Together
    // the two mints exercise every one of the 12 InvocationCaveats fields.
    @Test
    fun `hoursOfDay daysOfWeek inputSchema also round-trip via UCAN nb`() =
        runTest {
            assumeTrue(nativeAvailable, skipReason)

            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val admin = uniffi.scp.identityCreate("in_memory")
            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val member = uniffi.scp.identityCreate("in_memory")
            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val handle = uniffi.scp.contextCreate(admin, defaultParams(listOf("messages:read")))

            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val token =
                uniffi.scp.ucanMint(
                    handle,
                    member.did(),
                    listOf("messages:read"),
                    null,
                    companionCaveatsJson(),
                )

            val payloadNb =
                decodeJwtPayload(token.encoded())["nb"]?.jsonObject
                    ?: error("JWT payload missing `nb` field")
            assertCompanionNb(payloadNb)
        }

    // Mirrors `test_mint_limit_violation_surfaces_slug` in the Python
    // reference: 9 populated non-`origin_kind` fields exceeds
    // MAX_POPULATED_CAVEATS = 8 and must surface SCP-TOOL-6114 / the
    // `caveat-mint-limit-exceeded` slug.
    @Test
    fun `mint-limit violation surfaces caveat-mint-limit-exceeded slug (SCP-OUT-023 AC-6)`() =
        runTest {
            assumeTrue(nativeAvailable, skipReason)

            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val admin = uniffi.scp.identityCreate("in_memory")
            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val member = uniffi.scp.identityCreate("in_memory")
            // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
            val handle = uniffi.scp.contextCreate(admin, defaultParams(listOf("messages:read")))

            val thrown =
                runCatching {
                    // SCP-DEFAULT-INSTANCE-OK: raw UniFFI binding test; bypasses SDK facade by design
                    uniffi.scp.ucanMint(
                        handle,
                        member.did(),
                        listOf("messages:read"),
                        null,
                        overCapCaveatsJson(),
                    )
                }
            val error = thrown.exceptionOrNull()
            assertNotNull(error, "expected mint-limit violation to throw")
            val message = error.message.orEmpty() + " :: " + error.toString()
            assertTrue(
                message.contains("caveat-mint-limit-exceeded"),
                "expected error to carry caveat-mint-limit-exceeded slug, got: $message",
            )
        }
}

File: bindings/typescript/tests/caveats-roundtrip.test.ts
/**
 * SCP-OUT-023 AC-7 conformance: InvocationCaveats round-trip through the
 * real NAPI FFI bridge into the UCAN JWT `nb` field.
 *
 * Mirrors `bindings/python/tests/test_caveats_roundtrip.py`. The test:
 *
 *   1. Builds an `InvocationCaveats` object with all 12 fields populated
 *      (camelCase SDK shape).
 *   2. Calls `bridge.ucanMint(handle, did, caps, proofs, caveatsJson)` on
 *      the real native NAPI bridge.
 *   3. base64url-decodes the JWT payload segment (middle dot-segment).
 *   4. Asserts every wire-form caveat field appears in `payload.nb` with
 *      the expected value, and that fields the SDK omitted are NOT present
 *      in `nb` (SCP-OUT-018 `skip_serializing_if = Option::is_none`).
 *
 * The test is gated on the native NAPI bridge being loadable. When the
 * native addon or platform-specific package is not present, all tests skip
 * cleanly — mirroring `pytest.importorskip` in the Python conformance
 * test.
 *
 * Provenance:
 *   - .docs/prds/outlet.json — SCP-OUT-023 AC-7
 *   - .docs/specs/07-trust-validation-and-capabilities.md §7.3.8
 *   - bindings/python/tests/test_caveats_roundtrip.py (reference)
 */

import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import type { InvocationCaveats } from "../src/outlets";

// ---------------------------------------------------------------------------
// Native bridge availability guard — mirrors real-napi.test.ts.
// ---------------------------------------------------------------------------

type NativeBridge = Awaited<ReturnType<typeof import("../src/internal/bridge").getBridge>>;
type ServerAddon = {
  relayStartInMemory(): Promise<{
    readonly relayUrl: string;
    readonly relayPort: number;
    readonly isShutdown: boolean;
    shutdown(): void;
  }>;
  configureRelayTransport(relayUrl: string, localDid: string): Promise<void>;
};

let bridge: NativeBridge | null = null;
let serverAddon: ServerAddon | null = null;
let skipReason = "";

try {
  const { createNativeBridge } = await import("../src/internal/native.js");
  bridge = createNativeBridge();

  const { createRequire } = await import("node:module");
  const req = createRequire(import.meta.url);
  const platform = process.platform;
  const arch = process.arch;
  const platformMap: Record<string, string> = {
    "linux-x64": "@limn-works/scp-ts-napi-linux-x64-gnu",
    "linux-arm64": "@limn-works/scp-ts-napi-linux-arm64-gnu",
    "darwin-x64": "@limn-works/scp-ts-napi-darwin-x64",
    "darwin-arm64": "@limn-works/scp-ts-napi-darwin-arm64",
    "win32-x64": "@limn-works/scp-ts-napi-win32-x64-msvc",
  };
  const pkg = platformMap[`${platform}-${arch}`];
  if (pkg) {
    serverAddon = req(pkg) as ServerAddon;
  } else {
    skipReason = `No native addon for ${platform}-${arch}`;
  }
} catch (e: unknown) {
  const msg = e instanceof Error ? e.message : String(e);
  skipReason = `Native NAPI bridge not available: ${msg}`;
}

// ---------------------------------------------------------------------------
// Helpers.
// ---------------------------------------------------------------------------

/** base64url decode a string (with padding fixup) into a UTF-8 string. */
function base64urlDecode(input: string): string {
  // bun supports base64url natively via Buffer.from(..., "base64url").
  return Buffer.from(input, "base64url").toString("utf8");
}

/** Decode a UCAN JWT's payload (middle base64url segment) into an object. */
function decodeJwtPayload(encoded: string): Record<string, unknown> {
  const parts = encoded.split(".");
  if (parts.length !== 3) {
    throw new Error(`invalid JWT: expected 3 segments, got ${parts.length}`);
  }
  const payloadSegment = parts[1];
  if (payloadSegment === undefined || payloadSegment.length === 0) {
    throw new Error("JWT payload segment missing or empty");
  }
  return JSON.parse(base64urlDecode(payloadSegment)) as Record<string, unknown>;
}

/**
 * Serialize an `InvocationCaveats` into the bridge's wire JSON, mirroring
 * `caveatsToJson` in `bindings/typescript/src/ucan.ts`. Inlined here so the
 * test exercises the bridge contract directly without depending on the
 * SDK's mint wrapper indirection.
 */
function caveatsToWireJson(c: InvocationCaveats): string {
  const wire: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(c)) {
    if (value === undefined) continue;
    if (key === "rateWindow" && typeof value === "number") {
      wire[key] = { max: 1, windowSecs: value };
    } else {
      wire[key] = value;
    }
  }
  return JSON.stringify(wire);
}

// ---------------------------------------------------------------------------
// Test suite.
// ---------------------------------------------------------------------------

if (bridge === null || serverAddon === null) {
  describe("SCP-OUT-023 caveats round-trip (SKIPPED)", () => {
    test.skip(`real NAPI bridge unavailable: ${skipReason}`, () => {});
  });
} else {
  const napi = bridge;
  const addon = serverAddon;

  describe("SCP-OUT-023 AC-7: caveats round-trip via real NAPI", () => {
    let relayHandle: Awaited<ReturnType<typeof addon.relayStartInMemory>> | null = null;

    beforeAll(async () => {
      relayHandle = await addon.relayStartInMemory();
      // Bootstrap identity must precede configureRelayTransport so the
      // ContextManager OnceLock is initialised with a real DID.
      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const bootstrap = await napi.identityCreate("in_memory");
      await addon.configureRelayTransport(relayHandle.relayUrl, bootstrap.did);
    });

    afterAll(async () => {
      await napi.shutdown(1000);
      if (relayHandle && !relayHandle.isShutdown) {
        relayHandle.shutdown();
      }
    });

    test("all 12 InvocationCaveats fields survive marshal-unmarshal in JWT nb", async () => {
      // Cross-delegation: admin mints for member (avoids ADR-039 self-mint).
      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const admin = await napi.identityCreate("in_memory");
      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const member = await napi.identityCreate("in_memory");
      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const ctx = await napi.contextCreate(
        admin,
        JSON.stringify({
          ceiling: ["messages:read", "messages:write"],
          memoryScope: "ephemeral",
        }),
      );

      // §7.3.8 mint-limit: at most MAX_POPULATED_CAVEATS = 8 non-origin_kind
      // fields populated per envelope (origin_kind is structural and exempt).
      // We populate 8 budgeted fields covering both numeric ceilings and the
      // two list fields, plus origin_kind — the maximum mintable shape.
      const caveats: InvocationCaveats = {
        amountMaxPerCall: 100,
        amountMaxCumulative: 1_000,
        validFrom: 1_700_000_000,
        validUntil: 1_700_003_600,
        maxCalls: 42,
        rateWindow: 60, // 60s — wrapped into {max: 1, windowSecs: 60} on the wire
        // hoursOfDay / daysOfWeek / inputSchema deliberately omitted to stay
        // within the 8-populated-field cap; the absent-field assertions below
        // verify they do not appear in `nb`.
        allowedAdapters: ["native", "openai-compatible"],
        allowedTargetDids: ["did:dht:zMember", "did:dht:zOther"],
        originKind: "Action", // not counted toward the 8-field budget
      };

      const caveatsJson = caveatsToWireJson(caveats);

      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const token = await napi.ucanMint(
        ctx,
        member.did,
        ["messages:write"],
        undefined,
        caveatsJson,
      );

      expect(token.encoded).toBeTruthy();
      expect(typeof token.encoded).toBe("string");

      const payload = decodeJwtPayload(token.encoded);
      expect(payload).toHaveProperty("nb");
      const nb = payload["nb"] as Record<string, unknown>;
      expect(nb).toBeTruthy();

      // AC-7: every populated input field round-trips byte-for-byte.
      expect(nb["amountMaxPerCall"]).toBe(100);
      expect(nb["amountMaxCumulative"]).toBe(1_000);
      expect(nb["validFrom"]).toBe(1_700_000_000);
      expect(nb["validUntil"]).toBe(1_700_003_600);
      expect(nb["maxCalls"]).toBe(42);
      // rateWindow is normalized to the {max, windowSecs} wire shape.
      expect(nb["rateWindow"]).toEqual({ max: 1, windowSecs: 60 });
      expect(nb["allowedAdapters"]).toEqual(["native", "openai-compatible"]);
      expect(nb["allowedTargetDids"]).toEqual(["did:dht:zMember", "did:dht:zOther"]);
      expect(nb["originKind"]).toBe("Action");

      // Absent fields are omitted, never serialized as null
      // (SCP-OUT-018: skip_serializing_if = "Option::is_none").
      expect(nb).not.toHaveProperty("hoursOfDay");
      expect(nb).not.toHaveProperty("daysOfWeek");
      expect(nb).not.toHaveProperty("inputSchema");
    });

    test("hoursOfDay, daysOfWeek, inputSchema also round-trip via nb", async () => {
      // The 8-populated-field cap (§7.3.8 mint-limits) prevents covering all 12
      // fields in a single mint. This second mint covers the three fields the
      // primary test omitted, so the two tests together exercise every field.
      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const admin = await napi.identityCreate("in_memory");
      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const member = await napi.identityCreate("in_memory");
      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const ctx = await napi.contextCreate(admin, JSON.stringify({ ceiling: ["messages:read"] }));

      const caveats: InvocationCaveats = {
        hoursOfDay: 0x00ff_ffff, // 24-bit mask: every hour allowed
        daysOfWeek: 0x7f, // 7-bit mask: every day allowed
        inputSchema: {
          type: "object",
          properties: { x: { type: "number" } },
          required: ["x"],
        },
        originKind: "Query",
      };

      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const token = await napi.ucanMint(
        ctx,
        member.did,
        ["messages:read"],
        undefined,
        caveatsToWireJson(caveats),
      );
      const payload = decodeJwtPayload(token.encoded);
      const nb = payload["nb"] as Record<string, unknown>;
      expect(nb["hoursOfDay"]).toBe(0x00ff_ffff);
      expect(nb["daysOfWeek"]).toBe(0x7f);
      expect(nb["inputSchema"]).toEqual({
        type: "object",
        properties: { x: { type: "number" } },
        required: ["x"],
      });
      expect(nb["originKind"]).toBe("Query");

      // Fields the SDK did not populate must be omitted from `nb`.
      expect(nb).not.toHaveProperty("amountMaxPerCall");
      expect(nb).not.toHaveProperty("amountMaxCumulative");
      expect(nb).not.toHaveProperty("validFrom");
      expect(nb).not.toHaveProperty("validUntil");
      expect(nb).not.toHaveProperty("maxCalls");
      expect(nb).not.toHaveProperty("rateWindow");
      expect(nb).not.toHaveProperty("allowedAdapters");
      expect(nb).not.toHaveProperty("allowedTargetDids");
    });

    test("mint-limit violation surfaces SCP-TOOL-6114 slug (AC-6)", async () => {
      // Mirrors bindings/python/tests/test_caveats_roundtrip.py
      // ::test_mint_limit_violation_surfaces_slug. 9 populated non-origin
      // fields exceeds MAX_POPULATED_CAVEATS = 8.
      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const admin = await napi.identityCreate("in_memory");
      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const member = await napi.identityCreate("in_memory");
      // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
      const ctx = await napi.contextCreate(admin, JSON.stringify({ ceiling: ["messages:read"] }));

      const overCap: InvocationCaveats = {
        amountMaxPerCall: 1,
        amountMaxCumulative: 2,
        validFrom: 3,
        validUntil: 4,
        hoursOfDay: 0x00ff_ffff,
        daysOfWeek: 0x7f,
        maxCalls: 5,
        rateWindow: 60,
        inputSchema: { type: "object" }, // 9th populated field
      };

      await expect(
        // SCP-DEFAULT-INSTANCE-OK: raw NAPI bridge test; bypasses SDK facade by design
        napi.ucanMint(ctx, member.did, ["messages:read"], undefined, caveatsToWireJson(overCap)),
      ).rejects.toThrow(/caveat-mint-limit-exceeded/);
    });
  });
}
